src/lib/user_command_input/user_command_input.ml

```diff
@@ -0,0 +1,203 @@
+    | Signature of Signature.t
+    | Hd_index of Unsigned_extended.UInt32.t
+    | Keypair of Keypair.t
+  [@@deriving sexp, to_yojson]
+end
+
+[%%versioned
+module Stable = struct
+  module V1 = struct
+    type t =
+      ( Payload.Stable.V1.t
+      , Public_key.Compressed.Stable.V1.t
+      , Sign_choice.Stable.V1.t )
+      User_command.Poly.Stable.V1.t
+    [@@deriving sexp, to_yojson]
+
+    let to_latest = Fn.id
+  end
+end]
+
+type t = Stable.Latest.t [@@deriving sexp, to_yojson]
+
+let create ?(nonce_opt = None) ~fee ~valid_until ~memo ~body ~sender
+    ~sign_choice () : t =
+  let payload = Payload.create ~fee ~nonce_opt ~valid_until ~memo ~body in
+  {payload; sender; signature= sign_choice}
+
+let sign ~sender ~(user_command_payload : User_command_payload.t) = function
+  | Sign_choice.Signature signature ->
+      Option.value_map
+        ~default:(Deferred.return (Error "Invalid_signature"))
+        (User_command.create_with_signature_checked signature sender
+           user_command_payload)
+        ~f:Deferred.Result.return
+  | Keypair sender_kp ->
+      Deferred.Result.return (User_command.sign sender_kp user_command_payload)
+  | Hd_index hd_index ->
+      Secrets.Hardware_wallets.sign ~hd_index
+        ~public_key:(Public_key.decompress_exn sender)
+        ~user_command_payload
+
+let inferred_nonce ~f ~(sender : Public_key.Compressed.t) nonce_opt =
+  let open Result.Let_syntax in
+  let%bind txn_pool_or_account_nonce =
+    match Participating_state.active (f sender) |> Option.join with
+    | None ->
+        Error
+          "Couldn't infer nonce for transaction from specified `sender` since \
+           `sender` is not in the ledger or sent a transaction in transaction \
+           pool."
+    | Some nonce ->
+        Ok nonce
+  in
+  match nonce_opt with
+  | Some nonce ->
+      if Account_nonce.equal nonce txn_pool_or_account_nonce then Ok nonce
+      else
+        Error
+          (sprintf
+             !"Input nonce %s different from inferred nonce %s"
+             (Account_nonce.to_string nonce)
+             (Account_nonce.to_string txn_pool_or_account_nonce))
+  | None ->
+      Ok txn_pool_or_account_nonce
+
+let to_user_command ~result_cb ~infer_nonce ~logger uc_inputs :
+    (User_command.t list * (_ Or_error.t -> unit)) option Deferred.t =
+  let setup_user_command (client_input : t) :
+      User_command.t Deferred.Or_error.t =
+    (*create a user_command from input*)
+    let open Deferred.Or_error.Let_syntax in
+    let nonce_opt = client_input.payload.common.nonce in
+    let sender = client_input.sender in
+    let to_or_error : type a. (a, string) Result.t -> a Or_error.t =
+     fun e ->
+      let err_with_input str =
+        sprintf "Error creating user command: %s Error: %s"
+          (Yojson.Safe.to_string (to_yojson client_input))
+          str
+      in
+      Result.map_error e ~f:(Fn.compose Error.of_string err_with_input)
+    in
+    let%bind inferred_nonce =
```

Comment:
Won't this nonce be the same for every user command from the same sender?

Reply:
good catch, fixing it!